### DISCUSSION TITLE
Implicit updates getting in the way

### DISCUSSION THREAD

**Discussion by Randelung** [sqlalchemy/sqlalchemy#6544] [Usage Questions] [answered]
I have an application where I'm receiving lots of unconnected data, but sometimes the persist operation fails because one object implicitly adds more than just itself to a session. I have no idea what the issue could be.
I'm not sure I can recreate a minimal example demonstrating the issue since I'm not sure where to begin looking.
The data model consists of an aggregate (parent) that has datapoints (children) who in turn have instances of alerts (grandchildren). (There's more, but it shouldn't matter.) The aggregates don't change, the datapoints are frequently updated (but never inserted or deleted), while the amount of alerts grows (insert/update). The objects are linked using `selectin` and `immediate` loading.
The rows of data I'm receiving is from PLCs that communicate unique datapoints, which are then updated via the aggregate object and persisted into the database using a dedicated session.

```python
# in aggregate
async def set(self, value, datapoint_id):
  async with Session() as session:
    datapoint = self.get_datapoin(datapoint_id)
    datapoint.value = value
    session.add(datapoint)
    await session.commit()
```

In general, this works fine. The datapoint class is subclassed into several classes that need special treatment using a `polymorphic_identity`, but all of them just receive some additional attributes.
Except also some of them don't like to be persisted. Trying to update multiple datapoints on one aggregate works fine for some and throws an error for others. I've not found any pattern to this behavior. Tracking all incoming datapoints and lots of printing will show that pretty much all of the datapoints will have parallel open sessions - some of them without issues, others with an error saying the datapoint to be updated is already in a session from an unrelated datapoint.
I added a lock to each datapoint and around each session, which changed nothing. I'm thinking the session is pulling in the aggregate and all datapoints in some cases and just the one datapoint in others.

Is there a way to limit the added objects to just the explicitly added one? Or a different approach to avoid this issue?

```python
class Datapoint(Persistency.Base):
	__tablename__ = 'datapoints'

	id: int = Column('id', Integer, autoincrement=True, primary_key=True)
	role: str = Column('role', String(255))
	signalkennzeichnung: str = Column('signalkennzeichnung', String(255))
	model_id: int = Column('model', String(255), ForeignKey('models.id'))
	model: InformationModel = relationship('InformationModel', back_populates='datapoints', lazy='immediate')
	value = Column('value', String(255)) # all datapoints are string since most of them are an Enum.
	meldetext: str = Column('meldetext', String(255))
	guid: UUID = Column('guid', Persistency.GUID)
	suppressed: bool = Column('suppressed', Boolean)
	kind: str = Column('kind', CHAR(1)) # B(ebtrieblich) or O(perativ). Never O in VL.
	alerts: List[Alert] = relationship('Alert', back_populates='datapoint', lazy='selectin', primaryjoin='and_(Datapoint.id == Alert.datapoint_id, or_(Alert.time_went == None, Alert.time_acknowledged == None))')
	prio_pol: int = Column('prio_pol', TINYINT)
	prio_uhd: int = Column('prio_uhd', TINYINT)
	prio_vmz: int = Column('prio_vmz', TINYINT)
```

```python
class InformationModel(Persistency.Base): # shitty name for the aggregate
	__tablename__ = 'models'

	id: str = Column('id', String(255), primary_key=True)
	system: str = Column('system', String(255))
	aks: str = Column('aks', String(255))
	link: str = Column('link', String(255))
	datapoints: List[Datapoint] = relationship('Datapoint', back_populates='model', lazy='selectin')
	meldetext: str = Column('meldetext', String(255))
	bezeichnung: str = Column('bezeichnung', String(255))
	anzeigename: str = Column('anzeigename', String(255))
	guid: UUID = Column('guid', Persistency.GUID)
	type: str = Column('type', String(255))
	qs_id: int = Column('qs_id', Integer, ForeignKey('crosssections.id'))
	cross_section: CrossSection = relationship('CrossSection', back_populates='signals', lazy='immediate')
	qsk_id: int = Column('qsk_id', Integer, ForeignKey('qsk.id'))
	control_field_id: int = Column('controlfield', Integer, ForeignKey('controlfields.id'))

	__mapper_args__ = {
		'polymorphic_on': type,
		'polymorphic_identity': 'model',
		'with_polymorphic': '*'
	}

	async def set(self, role: Enum, value: Any) -> None:
		datapoint = self._check_role_validity_and_get_datapoint(role, value)
		if not datapoint:
			raise ValueError(f'{role} is not a valid role or {value} is not a valid value.')
		if role in self.roleEnums and value not in self.roleEnums[role]:
			raise ValueError(f'{value} is not a valid {role}.')
		if type(datapoint) is Datapoint:
			async with Persistency.Session() as session:
				datapoint.value = str(value)
				session.add(datapoint)
				await session.commit()
		else:
			self._pseudo_datapoints[role] = value
		if role in self.roleCallbacks:
			for callback in self.roleCallbacks[role]:
				await callback(self, role, value)
		send_model(self)
```

Example for a working subtype:
```python
class LEDSignal(InformationModel):
	__mapper_args__ = {
		'polymorphic_identity': 'ledsignal'
	}
```

Example of a failing subtype:
```python
class Hauptschalter(InformationModel):
	__mapper_args__ = {
		'polymorphic_identity': 'hauptschalter'
	}
```
Log goes about like this:
```
<incoming datapoint a.a>
<open session for datapoint a.a up to commit line>
<incoming datapoint a.b>
<open session for datapoint a.b up to commit line>
<incoming datapoint a.c>
<open session for datapoint a.c up to commit line>
<incoming datapoint b.a>
<open session for datapoint b.a up to commit line>
<incoming datapoint b.b>
'Object <Datapoing b.b> is already bound to session "4" (this is "5")'
<incoming datapoint b.c>
'Object <Datapoing b.c> is already bound to session "4" (this is "6")'
<incoming datapoint c.a>
<open session for datapoint c.a up to commit line>
<incoming datapoint c.b>
'Object <Datapoing b.c> is already bound to session "7" (this is "8")'
<incoming datapoint c.c>
'Object <Datapoing b.c> is already bound to session "7" (this is "9")'
<incoming datapoint d.a>
<open session for datapoint d.a up to commit line>
<incoming datapoint d.b>
<open session for datapoint d.b up to commit line>
<incoming datapoint d.c>
<open session for datapoint d.c up to commit line>
```
For a type of object, either all datapoints work fine, or all have the first-works-others-fail pattern.

I saw the `cascade` parameter for `relationship`, but I'm not sure what the default is or what I would need to change to not have the datapoints apparently go up to the parent and then down again.

Thanks for any input.

**Comment by zzzeek** [comment, marked as answer]
> I have an application where I'm receiving lots of unconnected data, but sometimes the persist operation fails because one object implicitly adds more than just itself to a session. I have no idea what the issue could be.

hi there -

it sounds like you're having a lot of problems but the descriptions here are too vague for us to give you any specific advice.    If you think objects are being added to the session that shouldn't be, you can debug this by implemententing a handler for the [before_attach](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_attach#sqlalchemy.orm.SessionEvents.before_attach) event listener, which will be triggered for every object added to the Session.    In here I would suggest some use of either print statements, pdb, or application-specific heuristics to identify the point at which additional objects are being added to the session that aren't wanted.


> I'm not sure I can recreate a minimal example demonstrating the issue since I'm not sure where to begin looking.

when you run your program, it fails.   keep removing things from the program until it stops failing; at that point, you will have identified at least one required factor in having it fail would be.    The most basic is to turn down the concurrency, see if the problem decreases with less concurrency, see if the problem goes away with no concurrency.   identifying the difference between when your program worked and when it stopped working is the main thing to isolate.

> The data model consists of an aggregate (parent) that has datapoints (children) who in turn have instances of alerts (grandchildren). (There's more, but it shouldn't matter.) The aggregates don't change, the datapoints are frequently updated (but never inserted or deleted), while the amount of alerts grows (insert/update). The objects are linked using `selectin` and `immediate` loading.

try not using "immediate" loading , not that there's anything wrong with immediate loading but perhaps it's a factor in what's going wrong, as this form of loading was almost never used prior to SQLAlchemy 1.4.

> Except also some of them don't like to be persisted. Trying to update multiple datapoints on one aggregate works fine for some and throws an error for others.

what kind of error?

>  I've not found any pattern to this behavior. Tracking all incoming datapoints and lots of printing will show that pretty much all of the datapoints will have parallel open sessions - some of them without issues, others with an error saying the datapoint to be updated is already in a session from an unrelated datapoint.
> I added a lock to each datapoint and around each session, which changed nothing. I'm thinking the session is pulling in the aggregate and all datapoints in some cases and just the one datapoint in others.

how would the Session be "pulling in" objects that you aren't passing to it?

> 
> Is there a way to limit the added objects to just the explicitly added one? 

if you take away [save-update](https://docs.sqlalchemy.org/en/14/orm/cascades.html?highlight=save%20update#save-update) cascade then absolutely nothing but the object you are adding using session.add() is added.    but use the before_attach event described above to trace what's happening at that level (I have a feeling this problem is data related, not ORM related).

> Or a different approach to avoid this issue?

without specific reproducing code, stack traces, etc., there's no way for me to have any idea what "the issue" is.



> 
> I saw the `cascade` parameter for `relationship`, but I'm not sure what the default is or what I would need to change to not have the datapoints apparently go up to the parent and then down again.

set the "cascade" parameter to blank on all relationship() calls:

```python
class Datapoint(Persistency.Base):

    # ...

	model: InformationModel = relationship('InformationModel', cascade='', back_populates='datapoints', lazy='immediate')
```

**Reply by Randelung** [reply]
Thanks so much for your quick reply.
I wasn't able to give you the actual outputs yesterday because of time constraints. I wrote the post at home without access to the production environment.
I'll try to get some insight today using `before_attach` and `cascade`.

**Reply by Randelung** [reply]
The `cascade=''` helped immediately.
When I tried it 2 minutes before I had to leave yesterday I only put it into one direction of the relationship. 😅
`befor_attach` confirmed that the aggregate was indeed being loaded implicitly.
Without any cascade I'm getting the desired behavior at the moment, so I'll mark this as solved. If I have more trouble, I'll ask in a new ticket.
Thanks!

**Reply by zzzeek** [reply]
OK good to know this made some progress, but I would note that turning off 'save-update' cascade is really unusual, im not sure if anyone has ever needed that before, so if you can provide small self-contained examples of what pattern you're using such that this is occurring that might shed light on some new use cases.

**Reply by Randelung** [reply]
Thanks for your concern. Sorry for not being more responsive at the moment, I've been on this basically every minute since the beginning of May. Not much wiggle room anymore for theoretical discussions, regardless of how useful they are.
I'm trying to create what you could almost call a realtime application for PLC data processing. The data model I manage is very static and using MariaDB system versioning, I'm only ever updating the datapoint values.  Some datapoints trigger alarms, which are Alarm rows and are just up to three timestamps (incoming, outgoing, acknowledged).
The database serves as historical storage for these datapoints and initial configuration for the application.
In order to be as realtime-y as possible I'm trying to have all of the configuration in RAM at initialisation - know all of the aggregates, datapoints and active alarms. Since the application is the only user for this database, the data ideally only ever flows into the database (only at init is it being read).
That plus asyncio made me believe that 'selectin' and 'immediate' (which you recommended a few weeks ago) would allow me to achieve this loading, and so far it's been almost working. However, the more I get infrequent errors about greenlets and detached instances, the more I fear that I might be using the library incorrectly or it's the wrong library for the use case. The application has a lifetime of about 10 years, so even infrequent errors are bound to occur.
I understand SQLAlchemy tries to be very transparent and loads entities when needed, which is ideal for large data sets, but I'm apparently trying to turn all of this off for a little more micromanage-y approach. 😅
I'd like to avoid one big session and the potential lazy loading if possible. selectin/immediate/session.refresh did that so far, and cascade='' was the next piece in the puzzle for parallel Datapoint updates. I'm trying to get a very static object landscape where only single entities are being written. ~~For example now I'm getting a DetachedInstanceError because I'm trying to traverse from a just updated Alarm to a Datapoint to an Aggregate, which it says is not in any session, even though the parent Datapoint and grandparent Aggregate are unchanged, even after a session.refresh of both.~~ Edit: Was idiot, had new VSCode without autosave enabled.

As the author of this library, you likely have much more experience than me when it comes to database paradigms - for me it's more of a side thought, which SQLAlchemy was supposed to take care of. The core really is just these three tables.
Aggregates 1:n Datapoints 1:n Alarms, where Datapoints.alarms is supposed to be only Alarms with one unfilled timestamp.
May I ask, just generally, how you would implement the entity loading? Maybe I'm already too far the rabbit hole to see the whole picture.

Thanks again.

**Reply by CaselIT** [reply]
If I understood correctly, your objective is to load some entity, let's call it A, once, at application loading, and then always use these entities, since they are static and never change. Then you have some connected entity, let's call it B, that are N:1 with A.

Given this, I don't see the point of using a relationship from A -> B. Directly load - save - update B, attaching the fk value manually. This would avoid having to re-attach the global A instance to a session each time you have to make a change to B. A can be expunged from the initial session after the first load and then it's just used as a "bag of properties".

Just my 2 cents here.

**Reply by Randelung** [reply]
Interesting idea. I would be able to keep the foreign keys, just not include any relationship statements, right? That would keep the db scheme unchanged, still enabling the usual create_all, and loading the tree structure would be trivial.
Thanks for your input, I'll give that a try.

**Reply by CaselIT** [reply]
Yes, the fk colum on B would still be there, you would just have to manage that manually, so when adding a B you have to explicitly set `B.a_fk_col = a.id` so that in the db is connected properly.

**Reply by Randelung** [reply]
Quick followup, I still have the alerts connected to the datapoint but disconnected the datapoint from the aggregate. The alerts relationship has a custom `primaryjoin='and_(Datapoint.id == Alert.datapoint_id, or_(Alert.time_went == None, Alert.time_acknowledged == None))'`, so updating a timestamp on an alert requires a session.refresh(datapoint).
~~Unless I'm misunderstanding something, that seems to wipe the object and any attributes I set, like the manual backref to the aggregate. Is there a way to either keep the backreference through a refresh or only refresh the relationship, but not the object itself?~~
I guess I could always store the reference before doing the refresh and just set it right afer.
Cheers.

Edit: The refresh isn't even the issue. I'm querying an Alarm by ID when acknowledging it and it loads a completely new object, which obviously doesn't have the aforementioned attributes. I'll have to change my API to follow the tree in that case.
Actually, that might be why I've been having troubles all this time. The new object would have to be replaced in all of its dependencies, prompting a rebuild of so many relationships.
I'll try to follow up again tomorrow, it's getting late.

Edit again: I didn't wait until tomorrow. I'm just iterating my tree to find the object, and upon finding the object I also find the backreference intact, even after a refresh. It's not the best solution, but if I need to improve performance, I'll make a dictionary with active alert IDs. On to the next thing!
